src/moin/utils/edit_locking.py

```diff
@@ -0,0 +1,331 @@
+            os.mkdir(os.path.join(app.cfg.instance_dir, PREVIEW))
+        con = sqlite3.connect(self.sql_filename)  # opens existing file or creates new file
+        cursor = con.cursor()
+        cursor.execute('''CREATE TABLE editlock(item_id TEXT PRIMARY KEY,
+                                                item_name TEXT,
+                                                user_name TEXT,
+                                                timeout FLOAT
+                                                )
+        ''')
+        cursor.execute('''
+            CREATE TABLE editdraft(user_name TEXT PRIMARY KEY,
+                                   item_id TEXT,
+                                   item_name TEXT,
+                                   rev_number INTEGER,
+                                   save_time INTEGER,
+                                   rev_id TEXT
+                                   )
+        ''')
+        con.commit()
+        return con
+
+    def make_draft_name(self, rev_id):
+        """Return a file name consisting of rev_id + user_name."""
+        keepchars = ('-', '.', '_')
+        return os.path.join(self.preview_path, rev_id + '-' + "".join(c for c in self.user_name if c.isalnum() or c in keepchars).rstrip())
+
+    def get_user_name(self):
+        """Return user name or user IP address."""
+        user_name = flaskg.user.name0
+        if user_name == ANON:
+            user_name = request.remote_user or request.remote_addr
+        return user_name
+
+    def put_draft(self, data_in, overwrite=True):
+        """
+        Only 1 item draft is saved per user. Most recent item draft overlays prior item draft.
+
+        If no editdraft row exists, an editdraft row is created when the user clicks the Modify
+        link. The row is updated when the user clicks Preview. The row is deleted when the
+        user clicks Cancel or Save/OK. The row is overwritten if the user abandons the edit
+        and clicks Modify for a different item.
+
+        The rev_number field is used to detect conflicting updates where first user's edit times
+        out, second user edits and saves, first user wakes up and does a save.
+        """
+        rev_id = self.rev_id
+        draft_rev_number = self.rev_number
+        draft, data = self.get_draft()
+        if draft:
+            # draft may be of this item or remnant of a prior abandoned edit
+            u_name, i_id, i_name, rev_number, save_time, i_rev_id = draft
+            if i_id == self.item_id:
+                if not overwrite:
+                    return
+                # in case of timeout and update by someone else get rev_number from prior Preview
+                draft_rev_number = rev_number  # XXX per line 1074 in __init__
+                rev_id = i_rev_id
+                self.draft_name = self.make_draft_name(rev_id)
+            self.cursor.execute('''DELETE FROM editdraft WHERE user_name = ? ''', (self.user_name, ))
+        if data_in:
```

Comment:
hmm, guess such kind of code is sqlite3 specific and if we run into troubles with it (like I did with the django app), changing to something else is not even simple (with the django ORM, I can also use pgsql - that's at least the theory).

Reply:
I also talked to some fellow python guys at last europython (or pycon.de) and they recommended always using a ORM and not directly talking to the DB implementiation (I asked about pg, but guess the same applies to sqlite3, maybe even more due to the scalability / parallel writers issues).

Reply:
SQLAlchemy supports an API between ORM and raw-queries.  For example:

```
metadata = sa.Metadata((bind=engine)
metadata.reflect()
t = sa.Table(table_name, self.metadata, autoload=True)
with engine.connect() as conn
    result = conn.execute(
        sa.select([t.c.id]).where(r.c.name == _name)
    )
```

I think this syntax is interpreted according for the underlying SQL engine used.

Reply:
See issue https://github.com/moinwiki/moin/issues/839

It has been ~7 years since I last wrote any sqlalchemy related code. I would welcome any help you could give.  I have no plans to work on https://github.com/moinwiki/moin/issues/839 soon. 

There are Locust load tests in /contrib/loadtesting/locust.